healthmanager/providers_health_manager.go

```diff
@@ -0,0 +1,126 @@
+package healthmanager
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	statusaggregator "github.com/status-im/status-go/healthmanager/aggregator"
+	"github.com/status-im/status-go/healthmanager/rpcstatus"
+)
+
+type ProvidersHealthManager struct {
+	mu          sync.RWMutex
+	chainID     uint64
+	aggregator  *statusaggregator.Aggregator
+	subscribers []chan struct{}
+}
+
+// NewProvidersHealthManager creates a new instance of ProvidersHealthManager with the given chain ID.
+func NewProvidersHealthManager(chainID uint64) *ProvidersHealthManager {
+	aggregator := statusaggregator.NewAggregator(fmt.Sprintf("%d", chainID))
+
+	return &ProvidersHealthManager{
+		chainID:    chainID,
+		aggregator: aggregator,
+	}
+}
+
+// Update processes a batch of provider call statuses, updates the aggregated status, and emits chain status changes if necessary.
+func (p *ProvidersHealthManager) Update(ctx context.Context, callStatuses []rpcstatus.RpcProviderCallStatus) {
+	p.mu.Lock()
+
+	previousStatus := p.aggregator.GetAggregatedStatus()
+
+	// Update the aggregator with the new provider statuses
+	for _, rpcCallStatus := range callStatuses {
+		providerStatus := rpcstatus.NewRpcProviderStatus(rpcCallStatus)
+		p.aggregator.Update(providerStatus)
+	}
+
+	newStatus := p.aggregator.GetAggregatedStatus()
+
+	shouldEmit := newStatus.Status != previousStatus.Status
+	p.mu.Unlock()
+
+	if !shouldEmit {
+		return
+	}
+
+	p.emitChainStatus(ctx)
+}
+
+// GetStatuses returns a copy of the current provider statuses.
+func (p *ProvidersHealthManager) GetStatuses() map[string]rpcstatus.ProviderStatus {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return p.aggregator.GetStatuses()
+}
+
+// Subscribe allows providers to receive notifications about changes.
+func (p *ProvidersHealthManager) Subscribe() chan struct{} {
+	p.mu.Lock()
```

Comment:
I think it shouldn't be a problem to protect the whole struct state with the same mutext, unless there is frequent access to `Update` and `Subscribe` methods at the same time.

Ideally we should not use 2 mutexes in the same struct to avoid unintentional deadlocks. Otherwise we have to ensure consistent locking order. Or I can use a `sync.Map` for subscribers